Qt: Implement graphics window

spycrab opened the pull request:
Screenshots:  
![General](https://i.imgur.com/bBXyojI.png)
![Enhancements](https://i.imgur.com/z7w6GEx.png)
![Hacks](https://i.imgur.com/ksaxSR0.png)
![Advanced](https://i.imgur.com/ADY8NGu.png)
TODO:
- [x] Fix the windows build
- [x] Implement blocking options when a game is running
- [x] Fix that certain options aren't disabled when they should be
- [x] Implement Software Rendering dialogue
- ~~Implement Shader Configuration Window (Low priority; Separate PR?)~~ (Currently not testable on my end).


Testing:
* ~~Test if every element has a description~~
* ~~Check if there are any mistakes in those descriptions~~
* ~~Check if all options behave as they should~~ (All do except two; Though seemingly not the fault of this PR, has to be investigated)
* ~~Check if you can crash Dolphin by changing certain settings~~

spycrab commented:
@ligfx @Starsam80 @MayImilae

ligfx commented:
Could you break this into separate commits, please? 1,700 LOC is a lot to review at once :(

spycrab commented:
@ligfx I would. Except for the fact that reverting past commits that are buried under commits you don't want to touch, is not something I have figured out yet.

Helios747 commented:
Kinda bugs me that we went with a sidebar tab system for Config, but a topbar tab system for Graphics.

Ultimately I don't really care enough for this to block this PR, but grr.

spycrab commented:
@Helios747 I discussed this on ``#dolphin-dev``: Although there is a plan to integrate it into the Settings Dialog later down the line, that's not happening right now as we want Qt to get up to speed. (Atleast that was the consensus).

lioncash commented on Source/Core/DolphinQt2/Config/Graphics/GraphicsWidget.h:
```diff
@@ -0,0 +1,39 @@
+public:
+  explicit GraphicsWidget(QWidget* parent);
+
+  bool eventFilter(QObject* obj, QEvent* event);
```
Missing override specifier

MayImilae commented:
Some small nitpicks:

The backend/fullscreen res/aspect ratio etc settings are divided a bit differently than WX. And eeeh, it matches the other windows more so I don't really care. Just noting it.

![blahs](https://user-images.githubusercontent.com/6551020/27206285-6071aaa8-51eb-11e7-9b2a-c04949be5d08.png)

The hacks tab Texture Cache setting is missing the "Accuracy:" line. It's kind of important to say what safe and fast are related to!

![texturecacheaccuracy](https://user-images.githubusercontent.com/6551020/27206047-82f6855a-51e9-11e7-819e-df3feb340c2c.png)

In the XFB section, please don't have XFB in the individual setting names, it's already stated at the top, and it's only there so users know what devs are talking about when they talk about XFB. Also, please place Virtual on the left, for the whole left is less right is more thing.

And if you are about to say that Texture Cache Accuracy has more accurate to the left... you're right! And it bugs me. It should be the other way around.



@spycrab @Helios747 I don't think graphics should be integrated with the main settings, honestly, because divisions like that really help keep it from being intimidating. But that's an argument for another time!

lioncash commented on Source/Core/DolphinQt2/MainWindow.cpp:
```diff
@@ -2,11 +2,12 @@
 #include <QIcon>
 #include <QMessageBox>
-
+#include <qpa/qplatformnativeinterface.h>
```
Qt [warns](https://github.com/qt/qtbase/blob/6bceb4a8a9292ce9f062a38d6fe143460b54370e/src/gui/kernel/qplatformnativeinterface.h#L47) about using this directly in applications.

spycrab replied:
It does warn about it, but as I see it there is no alternative with the current system.

lioncash commented on Source/Core/DolphinQt2/Config/Graphics/AdvancedWidget.cpp:
```diff
@@ -0,0 +1,183 @@
+void AdvancedWidget::AddDescriptions()
+{
+  static const QString TR_WIREFRAME_DESCRIPTION =
+      tr("Render the scene as a wireframe.\n\nIf unsure, leave this unchecked.");
```
I'm fairly certain you need to annotate these with [QT_TR_NOOP()](https://doc.qt.io/qt-5/qtglobal.html#QT_TR_NOOP) and use tr() at the usage site of them, in order for translations to work properly (see Qt's translation [guide](https://doc.qt.io/qt-5/i18n-source-translation.html)). Local statics are initialized on program startup, so runtime translation wouldn't work at the point of variable initialization.

lioncash commented on Source/Core/DolphinQt2/Config/Graphics/EnhancementsWidget.cpp:
```diff
@@ -0,0 +1,260 @@
+
+void EnhancementsWidget::AddDescriptions()
+{
+  static const QString TR_INTERNAL_RESOLUTION_DESCRIPTION =
```
Ditto about translations

lioncash commented on Source/Core/DolphinQt2/Config/Graphics/GeneralWidget.cpp:
```diff
@@ -0,0 +1,293 @@
+// We need QObject::tr
+#if defined(_WIN32)
+  static const QString TR_BACKEND_DESCRIPTION =
+      tr("Selects what graphics API to use internally.\nThe software renderer is extremely "
```
Ditto about translations

lioncash commented on Source/Core/DolphinQt2/Config/Graphics/GeneralWidget.h:
```diff
@@ -0,0 +1,58 @@
+  QCheckBox* m_show_fps;
+  QCheckBox* m_show_ping;
+  QCheckBox* m_log_rendertime;
+  QCheckBox* m_autoadjust_windowsize;
```
`m_autoadjust_window_size`

lioncash commented on Source/Core/DolphinQt2/Config/Graphics/GeneralWidget.h:
```diff
@@ -0,0 +1,58 @@
+  // Options
+  QCheckBox* m_show_fps;
+  QCheckBox* m_show_ping;
+  QCheckBox* m_log_rendertime;
```
`m_log_render_time`

lioncash commented on Source/Core/DolphinQt2/Config/Graphics/GeneralWidget.h:
```diff
@@ -0,0 +1,58 @@
+  QCheckBox* m_show_messages;
+  QCheckBox* m_keep_window_top;
+  QCheckBox* m_hide_cursor;
+  QCheckBox* m_render_mainwindow;
```
`m_render_main_window`

lioncash commented on Source/Core/DolphinQt2/Config/Graphics/HacksWidget.cpp:
```diff
@@ -0,0 +1,179 @@
+void HacksWidget::AddDescriptions()
+{
+  static const QString TR_SKIP_EFB_CPU_ACCESS_DESCRIPTION =
+      tr("Ignore any requests from the CPU to read from or write to the EFB.\nImproves "
```
Ditto about translations

ligfx commented on Source/Core/VideoCommon/VideoUtils.cpp:
```diff
@@ -0,0 +1,73 @@
+#include "VideoCommon/VideoUtils.h"
```
Missing copyright header.

ligfx commented on Source/Core/VideoCommon/VideoUtils.h:
```diff
@@ -0,0 +1,15 @@
+#pragma once
```
Missing copyright header

ligfx commented on Source/Core/VideoCommon/VideoUtils.h:
```diff
@@ -0,0 +1,15 @@
+
+namespace VideoUtils
+{
+std::vector<std::string> GetAvailableResolutions(X11Utils::XRRConfiguration* xrr_config);
```
- If there's no function definition when `APPLE` is defined, there probably shouldn't be a declaration either.
- This function should be in UICommon, not VideoCommon. Semantically, it has nothing to do with the video backends, and it's currently creating a cycle between the two modules (UICommon already depends on VideoCommon/VideoBackendBase).
- This shouldn't take an `XRRConfiguration` pointer when built without XRandR support. You could do something like:
  ```c++
  #ifdef _WIN32
    std::vector<std::string> GetAvailableResolutions();
  #elif defined(HAVE_XRANDR) && HAVE_XRANDR
    std::vector<std::string> GetAvailableResolutions(X11Utils::XRRConfiguration* xrr_config);
  #endif
  ```
  Alternatively, there's probably a way to have it take an opaque native window pointer (in WxWidgets, the result of `GetHandle()`) and internally figure out how to get an XRRConfiguration from that.

spycrab replied:
Those points seem reasonable.

ligfx replied:
Thanks, that looks better!

It doesn't look like the third point got fixed?

ligfx commented on Source/Core/VideoCommon/VideoUtils.h:
```diff
@@ -0,0 +1,15 @@
+namespace VideoUtils
+{
+std::vector<std::string> GetAvailableResolutions(X11Utils::XRRConfiguration* xrr_config);
+std::vector<std::string> GetAvailableAntialiasingModes(int& m_msaa_modes);
```
I don' think this function belongs in this commit.

spycrab replied:
Once moved to UICommon, I think it could.

ligfx replied:
I mean that it's not used anywhere in this commit, so it's out-of-place when reviewing :)

ligfx commented on Source/Core/DolphinQt2/Settings.cpp:
```diff
@@ -206,12 +206,22 @@ void Settings::SetStateSlot(int slot)
+  return SConfig::GetInstance().bRenderToMain;
+}
+
+void Settings::SetRenderToMain(bool rendertomain)
```
Why do all of these need to be in `Settings`, rather than accessed directly? Unless Settings is doing something special with them, like emitting signals, I don't see a need to add a layer of indirection here.

spycrab replied:
To have a wrapper around any ``SConfig`` code, so when the system is updated / changed only ``Settings.cpp`` is affected

ligfx replied:
This doesn't make sense to me. What problem are you trying to solve? If a configuration variable name is changed, a `find | sed -i s/bRenderToMain/myNewName/g` should fix DolphinQt2 just as well as the rest of Dolphin.

ZexaronS commented:
About textures, one small thing, keep in mind to make some space for the "Preload custom textures to VRAM" that I talked about with Stenzek, no I didn't mention to be an option at the time, but I can see it now that with bigger custom texture packs, it could be hitting over VRAM for some mid-to-lower end cards, certainly 3GB or lower of VRAM won't be enough for Hypatia's Zelda WW HD pack, which is at 3 GB (there's at least 150-300 MB VRAM overhead from browsers/desktop/etc)

ligfx commented on Source/Core/DolphinQt2/Config/Graphics/GraphicsWindow.h:
```diff
@@ -0,0 +1,33 @@
+{
+  Q_OBJECT
+public:
+  explicit GraphicsWindow(QWidget* parent, X11Utils::XRRConfiguration* xrr_config);
```
- The `QWidget* parent` argument should come at the end of the argument list, to match Qt.
- Does `XRRConfiguration` need to be passed in to this widget? Can it be created inside, or have the relevant information passed in directly?

spycrab replied:
1) Yes, good idea
2) 
* It can't be created inside; could have some unwanted side-effects AFAICT
* It probably has to be passed in as the ``GetAvailableResolutions()`` method needs it and it can't be constructed inside

Maybe there's a better way that I'm not seeing here, so I'm always open for suggestions.

ligfx replied:
What side-effects are you thinking of? Is there an issue with multiple displays, or something like that?

ligfx commented on Source/Core/DolphinQt2/Config/Graphics/GraphicsWindow.cpp:
```diff
@@ -0,0 +1,52 @@
+  main_layout->addWidget(m_tab_widget);
+  main_layout->addWidget(m_button_box);
+
+  auto* general_widget = new GeneralWidget(this, m_xrr_config);
```
These will be automatically parented by `addTab`, and shouldn't have their parent specified here.

http://doc.qt.io/qt-5/qtabwidget.html#details :
> The normal way to use QTabWidget is to do the following:
> [...]
> 2. Create a QWidget for each of the pages in the tab dialog, but do not specify parent widgets for them.

spycrab replied:
It's not about parenting (for most of the widgets atleast, for some the criticism is okay and can be changed), it's about being able to connect to the BackendChange signal of the Window

spycrab replied:
Okay nvm, should remove the ``QWidget(parent)`` thing everywhere, but still `` this`` has to be passed sometimes.

ligfx replied:
Widgets shouldn't require their parent to be anything more specific than a `QWidget*`, since that's intended solely for UI purposes. If something more specific needs to be passed in for non-UI purposes, that should be a separate argument.

In this case, couldn't you attach the signals here, rather than in the constructors? That would also simplify the widgets' internal logic by making it more "pure" (e.g. state changes pushed in, rather than pulled in).

ligfx commented on Source/Core/DolphinQt2/Config/Graphics/GraphicsWindow.cpp:
```diff
@@ -0,0 +1,52 @@
+  ConnectWidgets();
+
+  setWindowTitle(tr("Graphics"));
+  setWindowFlags(windowFlags() & ~Qt::WindowContextHelpButtonHint);
```
Could this be solved by making this not a `QDialog`, but a normal `QWidget` with the standalone-window flag?

spycrab replied:
Yes, but I don't see the advantage over using ``QDialog``, that makes it clear that this is in fact a Dialogue / Window

ligfx replied:
Mainly I was thinking that this isn't actually a dialog (none of these settings windows are). A dialog is intended for brief communication (questions, information, progress), not a full configuration window.

spycrab replied:
Still, I think a ``QDialog`` is still closer to a window than ``QWidget`` is.

ligfx replied:
The naming is a little confusing. A window is any `QWidget` without a parent, or with window flags `Qt::Window`. What's typically called a widget, is any `QWidget` that's not in the previous category.

spycrab replied:
Found a reason to keep using ``QDialog``: I need ``QDialogButtonBox``

ligfx replied:
`QDialogButtonBox` can be used anywhere, not only in `QDialogs` :)

ligfx commented on Source/Core/DolphinQt2/Config/Graphics/GraphicsSlider.h:
```diff
@@ -0,0 +1,23 @@
+// Copyright 2017 Dolphin Emulator Project
```
None of these should be new widget classes. They don't expose any methods, signals, or properties beyond their parent classes.

They can be more simply implemented with something like:

```c++
QComboBox* MakeGraphicsChoice(const QStringList& options, const Config::ConfigInfo<int>& setting)
{
  auto* combo_box = new QComboBox();

  combo_box->addItems(options);
  connect(combo_box, static_cast<void (QComboBox::*)(int)>(&QComboBox::currentIndexChanged),
          [&setting](int choice) { Config::SetBaseOrCurrent(setting, choice); });
  combo_box->setCurrentIndex(Config::Get(setting));

  if (Config::GetActiveLayerForConfig(setting) != Config::LayerType::Base)
  {
    QFont bf = combo_box->font();
    bf.setBold(true);
    combo_box->setFont(bf);
  }

  return combo_box;
}
```

spycrab replied:
I think this comes down to preferences, but I'd be fine implementing it this way, although not sure that's really necessary.

ligfx replied:
I wouldn't call it a preference! It's an engineering claim.

I see a number of advantages to doing it with pure functions:
- Better readability: A `GraphicsButton` (for example) is a class that a reader probably isn't familiar with. When reading code using it, they'll have to context switch, go to another file, and figure out what it does. This is a lot harder than skimming code that uses generic, well-known `QPushButtons` and sets attributes/signals in-line.

- Better semantics: The `GraphicsChoice` class (for example) would be more explicitly named `QComboBoxBoundToAConfigInfo`, which is a complete description of what it does. However, class usually aren't named with past-tense verbs, because verbs are mainly supposed to be for functions. Semantically, this is a combo box that has had some function applied to it, binding it to a config info.

- Better possibility for DRY/coalescing logic: The code for each of these widgets looks essentially like:
  ```c++
  connect(widget, &Widget::ValueChanged, [](new_value) {
    Config::SetBaseOrCurrent(setting, InvolutoryTransform(new_value));
  })
  setValue(InvolutoryTransform(Config::Get(setting));
  if (Config::GetActiveLayerForConfig(setting) != Config::LayerType::Base)
  {
    QFont bf = font();
    bf.setBold(true);
    setFont(bf);
  }
  ```

  With the appropriate setup (not saying I want you to do this, it's probably too complex to be worthwhile right now), this could be shared across all the needed widget types.

  It also makes it more obvious when shared logic is missing, and where there are potential opportunities for extracting repeated code: Why doesn't `GraphicsSlider` set its font to bold, when the others do? Because a slider widget doesn't come attached to a description label or value label. This seems like a Qt oversight, and there might be other code that could be made simpler with a combination-slider-attached-to-its-own-labels (for example, the Audio pane, when sliders need to enabled/disabled on the fly).

- Better separation of controllers and views: In the classical architecture, all control/model logic is separated from the UI elements. Obviously we're not adhering to this 100%, but it's a good idea to keep them separated as much as possible. The fewer places that conflate the two, the better.

I hope I can convince you 😄

spycrab replied:
Third opinion? @Lioncash

ligfx replied:
@spycrab Well, what do you think of the merits I described above? Do you not think those are good criteria for making design choices?

spycrab replied:
I wouldn't go so far as to say that those are bad criteria. But I disagree with how you interpret them. Even with your very detailed description of the possible advantages, none of it seems like it's a very big improvement even in the readability department  (or in some cases even no improvement at all, IMO).

In the end I don't want to spend my time changing functional and readable code that I - myself - think is perfectly serviceable. Not because I think I can claim some sort of huge objective merit (although I do think there is at least a small one) for that design decision, but because I think it's fine that way and perfectly harmless.

I could do a point by point response thingy, but I don't think that we are trying to figure out who is right on a mostly subjective matter and that would probably do no good.

I still do respect your opinion and thank you for taking the time to plough through my (sometimes massive) PRs.

Cheers.

ligfx commented on Source/Core/DolphinQt2/Config/Graphics/GraphicsWidget.h:
```diff
@@ -0,0 +1,39 @@
+class QGroupBox;
+class QLabel;
+
+class GraphicsWidget : public QWidget
```
I'm not a fan of this pattern which uses subclassing to share common helper functions. To me, subclassing means, you should be able to write generic logic that _consumes_ the base class, not generic logic that's used inside subclasses (and this isn't a mix-in).

Can the description box be turned into a standalone widget that each tab then adds at the bottom? That looks like it's basically how it's being used anyways. Maybe something like:

```c++
m_description_box = MakeDescriptionBox();
installEventFilter(m_description_box);
m_description_box->AddDescription(m_enable_wireframe, TR_WIREFRAME_DESCRIPTION);
```

spycrab replied:
Yeah sure, but (probably doesn't matter) why install an Event Filter on ``m_description_box`` did you mean ``m_enable_wireframe``?

spycrab replied:
Do you envision ``MakeDescriptionBox()`` to be part of ``GraphicsWidget``

spycrab replied:
Obsolete now.

ligfx replied:
What I envisioned was the description box as its own class, that each tab can add on the bottom (or the `GraphicsWindow` adds it to each, maybe?). Then this `GraphicsWidget` class would be removed entirely. Something like:

```c++
class ShowDescriptionsWhenHoveringOverOtherWidgetsBox : public QGroupBox
{
public:
  AddDescriptionForWidget(QWidget *widget, const char* description);
private:
  bool eventFilter(QObject* object, QEvent* event) override;
  QLabel* m_description;
  QHash<QObject*, const char*> m_widget_descriptions;
};
```

Can we remove this `GraphicsWidget` class entirely? The form layout shouldn't be used here, and exposing the descriptions`DescriptionAdded` could be done on each tab individually (either adding the signal on each class, or, probably more simply, exposing `void AddDescriptions(widget_implementing_descriptions)`).

The current version you uploaded places the description box slightly differently (outside the tabs box, screenshot attached) from how DolphinWX does it (inside the tabs box). Is that intended?

<img width="724" alt="screen shot 2017-06-18 at 2 56 07 pm" src="https://user-images.githubusercontent.com/594093/27264575-5cbe3ce4-5436-11e7-939e-c27885e5da13.png">

spycrab replied:
```
The current version you uploaded places the description box slightly differently (outside the tabs box, screenshot attached) from how DolphinWX does it (inside the tabs box). Is that intended?
```

ligfx replied:
Can we remove the `GraphicsWidget` class? It's not really doing anything.

MayImilae commented:
@ZexaronS Please keep the bikeshedding away from Qt work, we're trying to get what we have in WX copied over to Qt, and bikeshedding subjective UI elements is a deathkneel for UI PRs!

spycrab commented on Source/Core/DolphinQt2/MainWindow.cpp:
```diff
@@ -107,10 +115,20 @@ void MainWindow::CreateComponents()
+  m_graphics_window = new GraphicsWindow(
+      this,
+      new X11Utils::XRRConfiguration(
+          static_cast<Display*>(QGuiApplication::platformNativeInterface()->nativeResourceForWindow(
```
``XDisplayFromHandle`` seems to be ![unimplemented](https://github.com/dolphin-emu/dolphin/search?q=XDisplayFromHandle) (except for Wx)

ligfx replied:
That's my bad, carry on. (Though those functions shouldn't be declared in UICommon if they're not actually common!)

spycrab commented:
@MayImilae Implemented the changes you requested.

MerryMage commented on Source/Core/DolphinQt2/Config/Graphics/EnhancementsWidget.cpp:
```diff
@@ -0,0 +1,260 @@
+void EnhancementsWidget::SaveSettings()
+{
+  bool is_ssaa = m_aa_combo->currentText().startsWith(QStringLiteral("SSAA"));
+  g_Config.iMultisamples = m_aa_combo->currentIndex() - (is_ssaa ? m_msaa_modes : 0);
```
Config::SetBaseOrCurrent(Config::GFX_MSAA, /* ... */);

`g_Config` should be considered a read-only cache for the current configuration state (i.e.: a faster alternative to `Config::Get`, so `g_Config.iMultisamples` could be replaced with `Config::Get(Config::GFX_MSAA)`, for example).

spycrab replied:
Didn't know that was possible (as this was done the same way in Wx), but knowing that I'll use Config::SetBaseOrCurrent instead.

MerryMage commented on Source/Core/DolphinQt2/Config/Graphics/EnhancementsWidget.cpp:
```diff
@@ -0,0 +1,260 @@
+  bool is_ssaa = m_aa_combo->currentText().startsWith(QStringLiteral("SSAA"));
+  g_Config.iMultisamples = m_aa_combo->currentIndex() - (is_ssaa ? m_msaa_modes : 0);
+
+  g_Config.sPostProcessingShader = m_pp_effect->currentText().toStdString();
```
Config::SetBaseOrCurrent(Config::GFX_ENHANCE_POST_SHADER, /* ... */);

There do not be appear to be any other instances where you've directly written to `g_Config`.

spycrab replied:
See above.

MayImilae commented:
Some oddities I found while using this PR comparing WX and Qt under Windows!

![qtopengl](https://user-images.githubusercontent.com/6551020/27248018-2f94334c-52b3-11e7-97c5-27446ccbcd22.png)

On WX, OpenGL, Null, and Software doesn't have the adapter slot. But in Qt, OpenGL, Software and Null still have the adapter slot, but it's blank and is unable to expand.

Also, in Qt OpenGL the Fullscreen Resolution option is missing in Qt (but present in D3D and Vulkan), not sure why that is. Also, does anyone even set that to something other than Auto? ¯\_(ツ)_/¯

Oh, and super minor thing - the spacing between checkboxes is bigger in Qt. It's a super minor thing and totally not a blocker, but if that could be tightened up a little, that would be nice!

Another even more minor note in the image below: the description box in WX is always at the bottom, and allows any gaps to appear above it. In Qt, the Description box sticks to the bottom-most option, and allows a gap to fall beneath it. It might be better to go with the WX route in this, since it gives a universal place for the description box across tabs. Up to you!

Also, the software backend is very different, but you haven't implemented it yet, so moving on!

On to the next tab now...

![qt-enhancements](https://user-images.githubusercontent.com/6551020/27248700-42a915f2-52bc-11e7-9445-90e9dd69d003.png)

The Internal Resolution dropdown is missing the text label "for 1440p" for 4x native. Also, the resolution is wrong, it should be 2560x2112 but it is instead shown as 3200x2640, the same as 5x native. Whoops!

Qt doesn't have 2.5x native. Half resolutions _are_ trash, but, um... lots of users complain when we mess with those... I mean, most of our users are on terrible intel graphics, so... You may want to put it back in.

And Qt makes me scroll down the Internal Resolution list to see 8x Native. It's not a big deal, but it's kind of annoying, especially since I use 8x Native a fair bit when doing progress report work!

Another minor difference - items are missing the colon after their names. ...this is really pretty irrelevant, honestly, since "Name []" is just as clear as "Name: []", but you asked me to point out even minor things!

Also, something I noticed and haven't bothered to point out is that checkboxes are arranged a little differently, in 3 wide instead of 2 wide. In the Hacks tab it's 2 wide instead of one wide, etc. It's really really minor. There's also the different lengths for drop down items, but Qt's handling of that is kind of an improvement, so eh, don't care!

The hacks tab doesn't really have anything else to note, so moving to the Advanced Tab...

![d3dadvanced](https://user-images.githubusercontent.com/6551020/27248838-fe46eef0-52bd-11e7-9249-1f9223012e76.png)

In Qt, statistics in "Show Statistics" is not capitalized.

Prefetch Custom Texture is not greyed out at first, but when I cycle Load Custom Texture, then it is greyed out. Honestly WX is kind of unstable in this part too, but I'm pretty sure it should be greyed out when Load Custom Texture is unchecked, and selectable when Load Custom Texture is checked. It would also be nice if it unchecked itself before being greyed out, but I don't know if that's even possible.

Full Resolution Frame Dumps is greyed out and unselectable in D3D, but works in OpenGL and Vulkan. The option is just flat out missing with D3D in WX, since Stenzek never implemented it, so you may also want the option to be removed. As it is, it appears the option is on and I'm unable to turn off!

Borderless Fullscreen doesn't have a description on mouse over. I checked every single item, and that's the only one that's missing a description! Very nice!

Anyway, that's everything I found comparing the two in this PR. It's all pretty minor, which is pretty awesome! Keep up the good work!

spycrab commented:
> On WX, OpenGL, Null, and Software doesn't have the adapter slot. But in Qt, OpenGL, Software and Null   
> still have the adapter slot, but it's blank and is unable to expand.
> Also, in Qt OpenGL the Fullscreen Resolution option is missing in Qt (but present in D3D and Vulkan), not  
> sure why that is. Also, does anyone even set that to something other than Auto? ¯_(ツ)_/¯

Hopefully fixed.  

> Qt doesn't have 2.5x native. Half resolutions are trash, but, um... lots of users complain when we mess with those... I mean, most of our users are on terrible intel graphics, so... You may want to put it back in.

Added.

> In Qt, statistics in "Show Statistics" is not capitalized.

Fixed.

> Honestly WX is kind of unstable in this part too, but I'm pretty sure it should be greyed out when Load Custom Texture is unchecked, and selectable when Load Custom Texture is checked. It would also be nice if it unchecked itself before being greyed out, but I don't know if that's even possible.
> Full Resolution Frame Dumps is greyed out and unselectable in D3D, but works in OpenGL and Vulkan. The option is just flat out missing with D3D in WX, since Stenzek never implemented it, so you may also want the option to be removed. As it is, it appears the option is on and I'm unable to turn off!
> Borderless Fullscreen doesn't have a description on mouse over.

All Fixed.

> The Internal Resolution dropdown is missing the text label "for 1440p" for 4x native. Also, the resolution is wrong, it should be 2560x2112 but it is instead shown as 3200x2640, the same as 5x native. Whoops!

Oversaw this one, coming up next.

MayImilae commented:
Oh! I forgot to mention something! In WX, the backend is shown on top. "Dolphin OpenGL Graphics Configuration", etc. It's something that we put in to make it easier to diagnose screenshots that users give us. Are you planning to add this at some point?

spycrab commented:
@MayImilae Done.

spycrab commented:
@MayImilae Now the description box is always on the botttom

mimimi085181 commented:
So i found 3 "issues" on my 1st look:
- Full Resolution Frame Dumps is always greyed out for me, even with opengl or vulkan
- It's impossible to select any AA setting, the box is just empty
- The graphics adapter box is just empty. Can you grey that out for all backends that don't allow to choose one? Maybe with the text "default", so it doesn't look weird. Even on dx11, it doesn't allow me to select anything. Maybe it failts to detect this on my system? Windows 7 and an AMD Radeon HD 7790.

spycrab commented:
@mimimi085181 The first two should be fixed now.

mimimi085181 commented:
Thanks, but now Full Resolution Frame Dumps gets set to disabled when i switch to dx11. I guess that's related to you wanting to show the option as disabled when it's not working on that backend?

The stereoscopy settings are bugging me a bit. I can't select them when using dx11. Since i haven't used this, i don't know if that's correct. Anyways, how about greying out all the stereoscopy settings, when stereoscopy is set to off OR when stereoscopy is not supported? I know that's a bit out of scope since wx is not doing that, but i thought i could mention it anyways.

spycrab commented:
@mimimi085181 All fixed now AFAICT.

mimimi085181 commented:
Ok, other than the wireframe and progressive scan option not saving, i didn't find any more issues. In my opinion those 2 can be fixed later, especially the progressive scan, which might get changed anyways.

spycrab commented:
@mimimi085181 I agree, as these two are non-trivial to fix and it wouldn't do good fixing something that'll get changed soon anyway.

spycrab commented:
@dolphin-emu-bot rebuild

spycrab commented:
@dolphin-emu-bot rebuild

myownfriend commented:
I realize we want to first transfer the Wx layout to QT before we start talking about changes to the layout but Hide Mouse Cursor is already in the Interface area so it doesn't have to be in the Graphics settings.

MayImilae commented:
@myownfriend Leeeet's leave that sort of thing for after the move, please. It's arguably "graphics" since it's about what's on screen, so... it will be a discussion for sure.

myownfriend commented:
@MayImilae I understand. I just thought I'd bring it up in case no one was noticing.

spycrab commented:
@dolphin-emu-bot rebuild

spycrab commented:
Now in compliance with #5664.

shuffle2 commented:
rebase?

spycrab commented:
@shuffle2 Rebased and ready to roll.

shuffle2 commented:
Sorry looks like there's another conflict?

shuffle2 commented:
per irc there is some disagreement...

mimimi085181 commented:
Did you fix saving the wireframe option? It seems to be working now, and the info about it being broken could be removed. The only thing i don't like is the inconsistency on the spacing of the checkboxes, especially on the last tab. But that's not a blocker for me, it could be addressed later in another pr.

ligfx commented on Source/Core/DolphinQt2/Config/Graphics/GraphicsWidget.cpp:
```diff
@@ -0,0 +1,23 @@
+
+GraphicsWidget::GraphicsWidget(GraphicsWindow* parent)
+{
+  parent->RegisterWidget(this);
```
Having a child object reference the parent object like this increases coupling and is a bad pattern. Since it's being done to expose information during construction, one way to fix this would be to move the information out of the constructor. You could do something like:

```c++
class MyWidget
{
  GetDescriptions(std::function<void(QWidget*, const char*)> add_description)
  {
    add_description(child_widget, description);
    // ...
  }
};

class ParentWidget
{
  ParentWidget()
  {
    m_my_widget = new MyWidget();
    MyWidget->GetDescriptions(
        [=](QWidget* widget, const char* description) { OnDescriptionAdded(widget, description); });
  }
};
```

instead of the current

```c++
class MyWidget
{
  MyWidget(ParentWidget* parent)
  {
    parent->RegisterWidget(this);
    emit DescriptionAdded(child_widget, description);
    // ...
  }

signals:
  void DescriptionAdded(QWidget* widget, const char* description);
};

class ParentWidget
{
  ParentWidget() { m_my_widget = new MyWidget(this); }
  void RegisterWidget(MyWidget* my_widget)
  {
    connect(my_widget, &MyWidget::DescriptionAdded, this, &ParentWidget::OnDescriptionAdded);
  }
};
```

(which also has an additional benefits of being more direct and explicit).

spycrab replied:
This only complicates stuff IMO.

ligfx commented on Source/Core/DolphinQt2/Config/Graphics/GraphicsWidget.cpp:
```diff
@@ -0,0 +1,23 @@
+  parent->RegisterWidget(this);
+}
+
+void GraphicsWidget::AddDescription(QWidget* widget, const char* description)
```
Unneeded as `DescriptionAdded` can be called directly (`emit` is a blank macro).

ligfx commented on Source/Core/DolphinQt2/Config/Graphics/GeneralWidget.cpp:
```diff
@@ -0,0 +1,332 @@
+
+  // Video Section
+  auto* m_video_box = new QGroupBox(tr("Video"));
+  m_video_layout = new QFormLayout();
```
Using a `QFormLayout` here makes this look weird on macOS (and maybe KDE), and the non-form row at the bottom (V-Sync and Use Fullscreen) make it not be fully HIG compliant (it should be centered horizontally on macOS, looks neither macOS-style nor cross-platform-style as-is):

<img width="489" alt="screen shot 2017-06-27 at 12 59 05 pm" src="https://user-images.githubusercontent.com/594093/27607351-74bdfdf8-5b38-11e7-9d71-da159b7582cb.png">

Can you use a simple `QGridLayout`?

ligfx commented on Source/Core/DolphinQt2/Config/Graphics/GraphicsWindow.cpp:
```diff
@@ -0,0 +1,141 @@
+void GraphicsWindow::OnBackendChanged(const QString& backend)
+{
+  setWindowTitle(tr("Dolphin %1 Graphics Configuration").arg(backend));
+  if (backend == QStringLiteral("Software Renderer") && m_tab_widget->count() > 1)
```
Having this logic depend on the number of tab widgets currently visible seems fragile to me, could you make it depend directly on the old backend? Maybe an `m_current_backend_name`, or something like that?

spycrab replied:
I do not think this fragile at all.

spycrab commented on Source/Core/DolphinQt2/Config/Graphics/AdvancedWidget.cpp:
```diff
@@ -0,0 +1,207 @@
+
+void AdvancedWidget::OnBackendChanged()
+{
+  const auto supports_fr_framedumps = g_Config.backend_info.bSupportsInternalResolutionFrameDumps;
```
Nope. It being checked and it being supported are two entirely different things.

spycrab commented on Source/Core/DolphinQt2/Config/Graphics/EnhancementsWidget.cpp:
```diff
@@ -0,0 +1,305 @@
+    m_ir_combo->setCurrentIndex(m_ir_combo->count() - 1);
+  }
+
+  m_ir_combo->setMaxVisibleItems(m_ir_combo->count());
```
Leaving it as it is, as it is desired that all options are always visible.

spycrab commented on Source/Core/DolphinQt2/Config/Graphics/EnhancementsWidget.cpp:
```diff
@@ -0,0 +1,305 @@
+
+void EnhancementsWidget::LoadSettings()
+{
+  m_block_save = true;
```
It does. Loading the settings later on might trigger Saves that are not needed.

spycrab commented on Source/Core/DolphinQt2/Config/Graphics/GeneralWidget.cpp:
```diff
@@ -0,0 +1,327 @@
+
+void GeneralWidget::SaveSettings()
+{
+  // Video Backend
```
Nope.

sepalani commented on Source/Core/DolphinQt2/Config/Graphics/AdvancedWidget.cpp:
```diff
@@ -0,0 +1,208 @@
+// Copyright 2017 Dolphin5~ Emulator Project
```
"Dolphin" instead of "Dolphin5~".

sepalani commented on Source/Core/DolphinQt2/DolphinQt2.vcxproj:
```diff
@@ -1,4 +1,4 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
+<?xml version="1.0" encoding="utf-8"?>
```
Do not remove VS's BOM as I'm pretty sure it will be added again when saved by VS.

sepalani commented on Source/Core/VideoCommon/VideoCommon.vcxproj:
```diff
@@ -1,4 +1,4 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
+<?xml version="1.0" encoding="utf-8"?>
```
Ditto.

sepalani commented on Source/Core/VideoCommon/VideoCommon.vcxproj:
```diff
@@ -171,4 +171,4 @@
   </ImportGroup>
-</Project>
\ No newline at end of file
+</Project>
```
I'm pretty confident that VS will remove this newline.

shuffle2 replied:
Your comments are correct but this happens all the time in dolphin... Just ignore the extraneous back and forth changes🙂

BhaaLseN replied:
Or don't commit them at all, its not that hard :P

shuffle2 replied:
People that don't use vs are unreasonably against it.

BhaaLseN replied:
I meant not committing that line; not removing/untracking the project files.

shuffle2 replied:
I meant against dealing with it. Just look at our lint situation.

shuffle2 commented:
Lets just get this merged, and iterations on it can be done in successive prs? This is looking to become the victim of bike shedding.

sepalani commented on Source/Core/DolphinQt2/Config/Graphics/EnhancementsWidget.cpp:
```diff
@@ -0,0 +1,301 @@
+// Copyright 2015~7 Dolphin Emulator Project
```
What does the "~7" mean? From 2015 to 2017? Not sure that's the proper way to do so. Why not sticking with 2017?
